### PR TITLE
Eliminate allocation during HTTP2 path validation

### PR DESCRIPTION
**Pull request by JamesNK**
Showed up in profile:

image

Also is in HTTP/3 code and I changed it there. There seems to be a lot of copy and paste between HTTP/2 and 3. Is there a plan to improve that?

### CODE REVIEW

File: src/Servers/Kestrel/Core/src/Internal/Http2/Http2Stream.cs [hunk 1]
```diff
@@ -318,8 +318,8 @@ private bool TryValidatePath(ReadOnlySpan<char> pathSegment)
             try
             {
                 // The decoder operates only on raw bytes
-                var pathBuffer = new byte[pathSegment.Length].AsSpan();
-                for (int i = 0; i < pathSegment.Length; i++)
+                Span<byte> pathBuffer = stackalloc byte[pathSegment.Length];
```
Comment (gfoidl) [inline src/Servers/Kestrel/Core/src/Internal/Http2/Http2Stream.cs]: In general: it may be better to allocate to a constant size (and potentially slice to the correct size if needed), as the JIT can generate better code therefore.
This is especially true, if "skiplocalsinit = true", because then it's a just a (stack-) pointer move. For "skiplocalsinit = false" it's the pointer move + zeroing, but the logic to do so is still straight forward.

Comment (JamesNK) [reply src/Servers/Kestrel/Core/src/Internal/Http2/Http2Stream.cs]: The path segment could be any length, and changing this to process the path in chunks would be a non-trival change that I don't want to do right now..

Comment (JamesNK) [inline src/Servers/Kestrel/Core/src/Internal/Http2/Http2Stream.cs]: Should there be a maximum allowed stackalloc size? e.g.

```cs
Span<byte> pathBuffer = pathSegment.Length <= 1024 ? stackalloc byte[pathSegment.Length] : new byte[pathSegment.Length];
```

Comment (gfoidl) [reply src/Servers/Kestrel/Core/src/Internal/Http2/Http2Stream.cs]: Yes, especially when `pathSegment.Length` can be of any size, otherwise there may be a stackoverflow.

Perf-wise the JIT generates best code by using a constant:
```c#
Span<byte> pathBuffer = pathSegment.Length <= 1024
    ? stackalloc byte[1024].Slice(0, pathSegment.Length)
    : new byte[pathSegment.Length];
```
That's what I meant in my previous comment. It has nothing to do with processing in chunks. It's just about better / easier work for the JIT, as it knows how to stackalloc and doesn't need to emit logic for the allocation depending on the actual size.

`1024` seems pretty huge to me (and afaik this value isn't used in the runtime-repo), I would go with `256` (to be on the safe side with regards to SO-exceptions).

If `pathSegment.Length` will be quite often above 256, also use the array-pool:
```c#
byte[] bufferToReturnToPool = null;

Span<byte> pathBuffer = (pathSegment.Length <= 256
    ? stackalloc byte[256]
    : bufferToReturnToPool = ArrayPool<byte>.Shared.Rent(pathSegment.Length)
).Slice(0, pathSegment.Length);

try
{
    // ...
}
finally
{
    if (bufferToReturnToPool!=null)
    {
        bufferToReturnToPool.AsSpan(0, pathSegment.Length).Clear(); // optional
        ArrayPool<byte>.Shared.Return(bufferToReturnToPool);
    }
}
```

Comment (Tratcher) [reply src/Servers/Kestrel/Core/src/Internal/Http2/Http2Stream.cs]: The request has size limits, but they may still be larger than you'd want to stack alloc.

Comment (JamesNK) [reply src/Servers/Kestrel/Core/src/Internal/Http2/Http2Stream.cs]: I imagine that 256 will cover the vast majority of URLs

Comment (JamesNK) [reply src/Servers/Kestrel/Core/src/Internal/Http2/Http2Stream.cs]: Done

File: src/Servers/Kestrel/Core/src/Internal/Http2/Http2Stream.cs [hunk 2]
```diff
@@ -317,9 +317,16 @@ private bool TryValidatePath(ReadOnlySpan<char> pathSegment)
...
             {
+                const int MaxPathBufferStackAllocSize = 256;
+
                 // The decoder operates only on raw bytes
-                var pathBuffer = new byte[pathSegment.Length].AsSpan();
-                for (int i = 0; i < pathSegment.Length; i++)
+                Span<byte> pathBuffer = pathSegment.Length <= MaxPathBufferStackAllocSize
+                    // A constant size plus slice generates better code
+                    // https://github.com/dotnet/aspnetcore/pull/19273#discussion_r383159929
+                    ? stackalloc byte[MaxPathBufferStackAllocSize].Slice(0, pathSegment.Length)
```
Comment (halter73) [inline src/Servers/Kestrel/Core/src/Internal/Http2/Http2Stream.cs]: How does this compare to ` _context.MemoryPool.Rent(...`? With that, you could support pathSegments up to 4096 bytes without having allocate a new array.

Comment (JamesNK) [reply src/Servers/Kestrel/Core/src/Internal/Http2/Http2Stream.cs]: I don't know. I don't have experience using that type.

I could add a TODO comment to use it as a possible future optimization?

Comment (halter73) [reply src/Servers/Kestrel/Core/src/Internal/Http2/Http2Stream.cs]: FWIW it would look like the following:

```C#
try
{
    // The decoder operates only on raw bytes.
    IMemoryOwner<byte> memoryOwner;
    Span<byte> pathBuffer;

    // The MaxBufferSize for the default SlabMemoryPool is 4096 bytes.
    if (pathSegment.Length <= _context.MemoryPool.MaxBufferSize)
    {
        memoryOwner = _context.MemoryPool.Rent(pathSegment.Length);
        pathBuffer = memoryOwner.Memory.Span.Slice(0, pathSegment.Length);
    }
    else
    {
        memoryOwner = null;
        pathBuffer = new byte[pathSegment.Length];
    }

    using (memoryOwner)
    {
        for (int i = 0; i < pathSegment.Length; i++)
        {
            var ch = pathSegment[i];
            // ...
```

The pool is nice since it can give larger `Span<byte>`'s without worry of a stack overflow, but it isn't as efficient as stackallocing for small allocations. I ran a microbenchmark to see the difference for myself, and stackallocing performed almost 4 times better.

```c#
[Benchmark]
public void RentMemory()
{
    using (var owner = _slabMemoryPool.Rent(100))
    {
        owner.Memory.Span.Slice(0, 100);
    }
}

[Benchmark]
public void StackallocMemory()
{
    Span<byte> bytes = stackalloc byte[256].Slice(0, 100);
}
```

|           Method |      Mean |     Error |    StdDev |          Op/s | Allocated |
|----------------- |----------:|----------:|----------:|--------------:|----------:|
|       RentMemory | 28.604 ns | 0.5983 ns | 1.0788 ns |  34,959,914.1 |       0 B |
| StackallocMemory |  8.596 ns | 0.2041 ns | 0.2506 ns | 116,329,805.4 |       0 B |

Things flip when allocating 4000 bytes instead of 100 though:

|           Method |      Mean |     Error |   StdDev |         Op/s | Allocated |
|----------------- |----------:|----------:|---------:|-------------:|----------:|
|       RentMemory |  29.10 ns | 0.6079 ns | 1.433 ns | 34,359,941.0 |       0 B |
| StackallocMemory | 145.22 ns | 2.9209 ns | 5.116 ns |  6,885,903.9 |       0 B |

tl;dr: I think we should keep stackallocing since paths are generally pretty short. Maybe we can add a todo for using the pool when allocating between 256 and 3096 bytes.

**Review by halter73 — APPROVED** [review]

File: src/Servers/Kestrel/Core/src/Internal/Http2/Http2Stream.cs (hunk 2 above)
Comment (gfoidl) [reply src/Servers/Kestrel/Core/src/Internal/Http2/Http2Stream.cs]: > using the pool when allocating between 256 and 3096 bytes

The **3**096 is a typo? (`MemoryPool.MaxBufferSize` is 4096 as stated above, and for me just to understand why if it's not a typo).

> using the pool

Do you mean the Slabmemory-pool or the array-pool? The latter will allocate a new array on your behalf if it's to huge for the pool. That would safe another distinction within this code.
But as you said, if paths are generally pretty short, we should just keep the code as it is now.

Comment (halter73) [reply src/Servers/Kestrel/Core/src/Internal/Http2/Http2Stream.cs]: Yeah. 3096 was a typo. I was testing the SlabMemoryPool.
